jsonrpc/ipc.go

```diff
@@ -0,0 +1,221 @@
+	closing bool                  // closing indicates whether the server is in the process of shutting down.
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint and logger.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) *Ipc {
+	return &Ipc{
+		endpoint:       endpoint,
+		rpc:            rpc,
+		log:            log,
+		connParams:     DefaultIpcConnParams(),
+		conns:          make(map[net.Conn]struct{}),
+		events:         &SelectiveListener{},
+		socketCallback: func() {},
+	}
+}
+
+// WithSocketCallback sets the callback for socket creation.
+func (i *Ipc) WithSocketCallback(f func()) *Ipc {
+	i.socketCallback = f
+	return i
+}
+
+// WithConnParams sanity checks and applies the provided params.
+func (i *Ipc) WithConnParams(p *IpcConnParams) *Ipc {
+	i.connParams = *p
+	return i
+}
+
+// WithListener registers a NewRequestListener
+func (i *Ipc) WithListener(listener NewRequestListener) *Ipc {
+	i.events = listener
+	return i
+}
+
+// Run launches the IPC handler and handles any potential errors.
+// It is the caller's responsibility to cancel the provided context when they wish to
+// gracefully shut down the IPC handler.
+func (i *Ipc) Run(ctx context.Context) error {
+	if err := i.initialiseSocket(); err != nil {
+		return err
+	}
+
+	var (
+		errCh = make(chan error, 1)
+		wg    conc.WaitGroup
+	)
+	defer wg.Wait()
+	wg.Go(func() { errCh <- i.run(ctx) })
+
+	select {
+	case <-ctx.Done():
+	case err := <-errCh:
+		if !isSocketError(err) {
+			i.log.Warnw("Listener failure", "err", err)
+		}
+	}
+	return i.stop()
+}
+
+// initialiseSocket initialises the socket and executes the socket callback upon successful creation.
+func (i *Ipc) initialiseSocket() error {
+	listener, err := createListener(i.endpoint)
+	if err != nil {
+		return err
+	}
+	i.listener = listener
+	i.socketCallback()
+	return nil
+}
+
+// run continuously accepts and serves incoming connections.
+func (i *Ipc) run(ctx context.Context) error {
+	for {
+		conn, err := i.listener.Accept()
+		if netutil.IsTemporaryError(err) {
+			i.log.Errorw("Failed to accept connection", "err", err)
+			continue
+		} else if err != nil {
+			return err
+		}
+		i.connWg.Add(1)
+		go i.serveConn(ctx, newIpcConn(conn, i.connParams))
+	}
+}
+
+// setupConn adds a connection and checks for handler closing.
+func (i *Ipc) setupConn(conn net.Conn) bool {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.conns[conn] = struct{}{}
+	return !i.closing
+}
+
+// cleanupConn calls cleanupConnNoLock with explicit lock
+func (i *Ipc) cleanupConn(conn net.Conn) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.cleanupConnNoLock(conn)
+}
+
+// cleanupConnNoLock frees resources
+func (i *Ipc) cleanupConnNoLock(conn net.Conn) {
+	_, ok := i.conns[conn]
+	delete(i.conns, conn)
+	if ok {
+		conn.Close()
+		i.connWg.Done()
+	}
+}
+
+// serveConn handles incoming connection.
+func (i *Ipc) serveConn(ctx context.Context, conn net.Conn) {
+	defer i.cleanupConn(conn)
```

Comment:
```suggestion
	defer func() {
	        i.mu.Lock()
	        defer i.mu.Unlock()
	        i.cleanupConnNoLock(conn)
	}()
```
Let's remove some indirection.